### PR TITLE
Feature/stop dp dimension extractor killing itself

### PR DESCRIPTION
**Pull request by DavidSubiros**
### What

Now the service doesn't die if external services don't run, or if validation fails. If services die during this service executions, it does not die either.

### How to review

- Check that you can run the service without Kafka, and it does not die (it logs the errors).
- Try to start the service with Kafka running, then kill Kafka, and the service does not die (it logs the errors).
- Try to start the service with an invalid token or no token, and it should not die.
- The service should continue to function as expected when the rest of the services run.

### Who can review

Anyone

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -114,60 +112,99 @@ func main() {
...
-		log.Debug("closing dimension extracted kafka producer", nil)
-		service.DimensionExtractedProducer.Close(ctx)
-		log.Debug("closed dimension extracted kafka producer", nil)
 
-		log.Debug("closing down dimension extracted error producer", nil)
-		dimensionExtractedErrProducer.Close(ctx)
-		log.Debug("closed dimension extracted error producer", nil)
+		// If DimensionExtracted kafka producer exists, close it
+		if serviceList.DimensionExtractedProducer {
+			log.Debug("closing dimension extracted kafka producer", nil)
```
Comment (daiLlew) [inline main.go]: Minor point - I know you wouldn't be aware of this but we have a logging standard where we don't have dynamic or variable dependant messages. 
You have the multiple messages for the same event but the only thing that is different is the producer being closed. If you make the message more general you can move the specific context to the logging params i.e.

```
log.Debug("closing kafka producer", log.Data{"producer": dimension extracted})
```

File: initialise/initialise.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+// GetProducer returns a kafka producer
+func (e *ExternalServiceList) GetProducer(kafkaBrokers []string, topic, name string, envMax int) (kafkaProducer kafka.Producer, err error) {
+	kafkaProducer, err = kafka.NewProducer(kafkaBrokers, topic, envMax)
+	if err == nil {
+		switch {
+		case name == DimensionExtractedProducer:
+			e.DimensionExtractedProducer = true
+		case name == DimensionExtractedErrProducer:
+			e.DimensionExtractedErrProducer = true
```
[the remark this answers is not in the document]
Comment (DavidSubiros) [reply initialise/initialise.go]: Maybe I can add the name provided and the valid names to make the error more useful?

File: initialise/initialise.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+		e.Vault = true
+	}
+
+	return
+}
+
+// GetAwsSession returns an AWS client for the AWS region provided in Config
+func (e *ExternalServiceList) GetAwsSession(cfg *config.Config) (awsSession *session.Session, err error) {
+	awsSession, err = session.NewSession(&aws.Config{Region: &cfg.AWSRegion})
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (DavidSubiros) [reply initialise/initialise.go]: yes, I think this was changed in another commit, thanks!

File: initialise/initialise.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (DavidSubiros) [reply initialise/initialise.go]: I think I will use an iota for the const, to make the code more elegant and easier to maintain.

File: initialise/initialise.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (DavidSubiros) [reply initialise/initialise.go]: Instead of this, we decided to handle the err != nil case inside the if for all the cases.

File: main.go [hunk 4]
```diff
@@ -114,60 +112,99 @@ func main() {
...
+	// Log non-fatal errors, without exiting
+	go func() {
+		var consumerErrors chan (error)
+		if serviceList.Consumer {
+			consumerErrors = syncConsumerGroup.Errors()
+		} else {
+			consumerErrors = make(chan error, 1)
+		}
+
+		var producerErrors chan (error)
```
Comment (nshumoogum) [inline main.go]: You could initialise both channels on the same line like so:

`var consumerErrors, producerErrors chan (error)`

File: main.go [hunk 5]
```diff
@@ -114,60 +112,99 @@ func main() {
...
+			log.ErrorC("kafka producer", producerError, nil)
+		case apiError := <-apiErrors:
+			log.ErrorC("server error", apiError, nil)
+		}
+	}()
 
-	// block until a fatal error, signal or eventLoopDone - then proceed to shutdown
+	// Block until a fatal error occurs
 	select {
 	case <-eventLoopDone:
```
Comment (nshumoogum) [inline main.go]: Should we not remove this case to prevent the app proceeding to shutdown early?

Comment (DavidSubiros) [reply main.go]: Done

**Pull request merged**